File: web/src/shared/realtime/protocol.ts
import type {
  CallEndReason,
  CallMediaType,
  ConversationMember,
  Message,
  UserSummary,
} from "@/shared/api/types";

export type ClientCommandType =
  | "message.send"
  | "direct_message.send"
  | "conversation.read"
  | "heartbeat.ping"
  | "call.invite"
  | "call.accept"
  | "call.connected"
  | "call.reject"
  | "call.cancel"
  | "call.hangup"
  | "call.signal";

export type MessageSendPayload = {
  conversation_id: string;
  client_msg_id: string;
  body: string;
};

type DirectMessageTarget =
  | { target_username: string; target_user_id?: never }
  | { target_username?: never; target_user_id: string };

export type DirectMessageSendPayload = DirectMessageTarget & {
  client_msg_id: string;
  body: string;
};

export type ConversationReadPayload = {
  conversation_id: string;
  read_seq: number;
};

export type HeartbeatPingPayload = {
  client_time: string;
};

export type CallInvitePayload = {
  conversation_id: string;
  media_type: CallMediaType;
};

export type CallIdPayload = {
  call_id: string;
};

export type CallHangupPayload = {
  call_id: string;
  reason?: CallEndReason;
};

export type CallSignalPayload = {
  call_id: string;
  signal_type: "offer" | "answer" | "ice_candidate";
  data: unknown;
};

export type ClientCommandPayloadByType = {
  "message.send": MessageSendPayload;
  "direct_message.send": DirectMessageSendPayload;
  "conversation.read": ConversationReadPayload;
  "heartbeat.ping": HeartbeatPingPayload;
  "call.invite": CallInvitePayload;
  "call.accept": CallIdPayload;
  "call.connected": CallIdPayload;
  "call.reject": CallIdPayload;
  "call.cancel": CallIdPayload;
  "call.hangup": CallHangupPayload;
  "call.signal": CallSignalPayload;
};

export type RealtimeOutgoing<
  TType extends ClientCommandType = ClientCommandType,
  TPayload = ClientCommandPayloadByType[TType],
> = {
  id?: string;
  type: TType;
  payload: TPayload;
};

export type SendMessageResult = {
  conversation_id: string;
  message: Message;
};

export type ConversationReadResult = {
  conversation_id: string;
  read_seq: number;
};

export type HeartbeatPongPayload = {
  server_time: string;
};

export type RealtimeErrorPayload = {
  code: string;
  message: string;
};

export type CallSummary = {
  call_id: string;
  conversation_id: string;
  caller: UserSummary;
  callee: UserSummary;
  caller_client_id: string;
  accepted_client_id: string | null;
  media_type: "audio" | "video";
  state: "ringing" | "connecting" | "active" | "ended";
  started_at: string;
  accepted_at: string | null;
  ended_at: string | null;
  end_reason: CallEndReason | null;
};

export type MessageCreatedEvent = {
  type: "message.created";
  payload: {
    conversation_id: string;
    message: Message;
  };
};

export type ConversationReadUpdatedEvent = {
  type: "conversation.read_updated";
  payload: {
    conversation_id: string;
    user_id: string;
    read_seq: number;
  };
};

export type ConversationMemberAddedEvent = {
  type: "conversation.member_added";
  payload: {
    conversation_id: string;
    member: ConversationMember;
  };
};

export type ConversationMemberLeftEvent = {
  type: "conversation.member_left";
  payload: {
    conversation_id: string;
    user_id: string;
  };
};

export type ConversationDissolvedEvent = {
  type: "conversation.dissolved";
  payload: {
    conversation_id: string;
    user_id: string;
  };
};

export type CallStateEventType =
  | "call.incoming"
  | "call.ringing"
  | "call.accepted"
  | "call.connected"
  | "call.rejected"
  | "call.canceled"
  | "call.ended"
  | "call.busy";

export type CallStateEvent = {
  type: CallStateEventType;
  payload: {
    call: CallSummary;
  };
};

export type CallSignalEvent = {
  type: "call.signal";
  payload: CallSignalPayload;
};

export type ServerDrainingEvent = {
  type: "server.draining";
  payload?: {
    retry_after_ms?: number;
  };
};

export type RealtimeServerEvent =
  | MessageCreatedEvent
  | ConversationReadUpdatedEvent
  | ConversationMemberAddedEvent
  | ConversationMemberLeftEvent
  | ConversationDissolvedEvent
  | CallStateEvent
  | CallSignalEvent
  | ServerDrainingEvent;

export type CommandOkEnvelope<TPayload = unknown> = {
  id?: string;
  type: `${string}.ok`;
  payload: TPayload;
};

export type HeartbeatPongEnvelope = {
  id?: string;
  type: "heartbeat.pong";
  payload: HeartbeatPongPayload;
};

export type RealtimeErrorEnvelope = {
  id?: string;
  type: "error";
  error: RealtimeErrorPayload;
};

export type RealtimeIncoming =
  | RealtimeServerEvent
  | CommandOkEnvelope
  | HeartbeatPongEnvelope
  | RealtimeErrorEnvelope;

File: src/ws/protocol.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use uuid::Uuid;

use crate::{error::ErrorCode, ids::UserId};

#[derive(Debug, Clone, PartialEq, Deserialize)]
pub struct ClientEnvelope {
    #[serde(default)]
    pub id: Option<String>,
    #[serde(rename = "type")]
    pub message_type: String,
    #[serde(default)]
    pub payload: Value,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ServerEnvelope {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<String>,
    #[serde(rename = "type")]
    pub message_type: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub payload: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<ServerError>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ServerError {
    pub code: String,
    pub message: String,
}

impl ServerEnvelope {
    pub fn ok(id: Option<String>, message_type: impl Into<String>, payload: Value) -> Self {
        Self {
            id,
            message_type: message_type.into(),
            payload: Some(payload),
            error: None,
        }
    }

    pub fn event(message_type: impl Into<String>, payload: Value) -> Self {
        Self {
            id: None,
            message_type: message_type.into(),
            payload: Some(payload),
            error: None,
        }
    }

    pub fn error(id: Option<String>, code: ErrorCode, message: impl Into<String>) -> Self {
        Self {
            id,
            message_type: "error".to_string(),
            payload: None,
            error: Some(ServerError {
                code: code.as_str().to_string(),
                message: message.into(),
            }),
        }
    }
}

#[derive(Debug, Clone, Deserialize)]
pub struct MessageSendPayload {
    pub conversation_id: Uuid,
    pub client_msg_id: String,
    pub body: String,
}

#[derive(Debug, Clone, Deserialize)]
pub struct DirectMessageSendPayload {
    #[serde(default)]
    pub target_user_id: Option<UserId>,
    #[serde(default)]
    pub target_username: Option<String>,
    pub client_msg_id: String,
    pub body: String,
}

#[derive(Debug, Clone, Deserialize)]
pub struct ConversationReadPayload {
    pub conversation_id: Uuid,
    pub read_seq: i64,
}

#[derive(Debug, Clone, Deserialize)]
pub struct HeartbeatPingPayload {
    #[serde(default)]
    pub client_time: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct CallInvitePayload {
    pub conversation_id: Uuid,
    pub media_type: crate::calls::types::CallMediaType,
}

#[derive(Debug, Clone, Deserialize)]
pub struct CallIdPayload {
    pub call_id: Uuid,
}

#[derive(Debug, Clone, Deserialize)]
pub struct CallHangupPayload {
    pub call_id: Uuid,
    #[serde(default)]
    pub reason: Option<crate::calls::types::CallEndReason>,
}

#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum CallSignalType {
    Offer,
    Answer,
    IceCandidate,
}

#[derive(Debug, Clone, Deserialize)]
pub struct CallSignalPayload {
    pub call_id: Uuid,
    pub signal_type: CallSignalType,
    pub data: serde_json::Value,
}
